server/server.go

```diff
@@ -2187,7 +2189,7 @@ func findBestQueueSub(sl []*subState) (rsub *subState) {
 
 		// Favor non stalled subscribers and clients that do not have
 		// failed heartbeats
-		if (!sStalled || rStalled) && (!sHasFailedHB || rHasFailedHB) && (sOut < rOut) {
+		if (!sStalled && !sHasFailedHB && (rStalled || rHasFailedHB)) || (sOut < rOut) {
```

Comment:
This logic takes a bit to grok and got me thinking about the func as a whole. Could this be simplified to make the criteria evaluation for "BestQueueSub" more clear and reduce locking:

```
// ... (line 2170)
leastOutstanding := int(^uint(0) >> 1) // set to max int
for _, sub := range sl {

	sub.RLock()
	sOut := len(sub.acksPending)
	sStalled := sub.stalled
	sHasFailedHB := sub.hasFailedHB
	sub.RUnlock()

	if !sStalled && !sHasFailedHB {
		if sOut < leastOutstanding {
			leastOutstanding = sOut			
			rsub = sub
		}
	}
}
...
```

I realize this would increase the time between evaluating a subscriber's state and possibly sending a message to it, but it would also reduce the locks so hopefully the loop would run a tiny bit faster.